// react-google-auth/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.GetCurrentAuthResponse = exports.GetCurrentUser = exports.GetAuthInstance = exports.Compose = exports.GoogleAuthHock = exports.AsyncLoaderHock = exports.ScriptLoaderHock = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ScriptLoaderHock = require('./ScriptLoaderHock');

var _ScriptLoaderHock2 = _interopRequireDefault(_ScriptLoaderHock);

var _AsyncLoaderHock = require('./AsyncLoaderHock');

var _AsyncLoaderHock2 = _interopRequireDefault(_AsyncLoaderHock);

var _GoogleAuthHock = require('./GoogleAuthHock');

var _GoogleAuthHock2 = _interopRequireDefault(_GoogleAuthHock);

var _Compose = require('./Compose');

var _Compose2 = _interopRequireDefault(_Compose);

var _Utils = require('./Utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (config) {
    var clientId = config.clientId,
        discoveryDocs = config.discoveryDocs,
        _config$loader = config.loader,
        loader = _config$loader === undefined ? function () {
        return _react2.default.createElement(
            'div',
            null,
            'Loading'
        );
    } : _config$loader,
        scope = config.scope,
        _config$signIn = config.signIn,
        signIn = _config$signIn === undefined ? function (_ref) {
        var onSignInClick = _ref.onSignInClick;
        return _react2.default.createElement(
            'button',
            { onClick: onSignInClick },
            'Sign in'
        );
    } : _config$signIn;


    return (0, _Compose2.default)((0, _ScriptLoaderHock2.default)({
        scripts: ['https://apis.google.com/js/api.js'],
        loader: loader
    }), (0, _AsyncLoaderHock2.default)({
        callbackOnMount: function callbackOnMount(onComplete) {
            return gapi.load('client:auth2', onComplete);
        },
        loader: loader
    }), (0, _GoogleAuthHock2.default)({
        clientId: clientId,
        discoveryDocs: discoveryDocs,
        scope: scope,
        signIn: signIn
    }));
};

exports.ScriptLoaderHock = _ScriptLoaderHock2.default;
exports.AsyncLoaderHock = _AsyncLoaderHock2.default;
exports.GoogleAuthHock = _GoogleAuthHock2.default;
exports.Compose = _Compose2.default;
exports.GetAuthInstance = _Utils.GetAuthInstance;
exports.GetCurrentUser = _Utils.GetCurrentUser;
exports.GetCurrentAuthResponse = _Utils.GetCurrentAuthResponse;
